src/__experimental__/components/date/date.component.js

```diff
@@ -110,48 +105,72 @@ class BaseDateInput extends React.Component {
   openDatePicker = () => {
     this.isBlurBlocked = true;
     document.addEventListener('click', this.closeDatePicker);
-    this.updateSelectedDate(this.props.value);
+    this.updateSelectedDate(this.props.value || isoFormattedValueString(this.state.visibleValue));
     this.setState({ isDatePickerOpen: true });
   };
 
-  reformatVisibleDate = () => {
-    const { visibleValue } = this.state;
 
-    if (DateHelper.isValidDate(visibleValue)) {
-      this.setState({ visibleValue: formatDateToCurrentLocale(visibleValue) });
-    }
+  updateValidEventValues = (value) => {
+    this.setState({
+      visibleValue: formatDateToCurrentLocale(value),
+      lastValidEventValues: {
+        formattedValue: formatDateToCurrentLocale(value),
+        rawValue: isoFormattedValueString(value)
+      }
+    });
   }
 
+   reformatVisibleDate = () => {
+     const { visibleValue } = this.state;
+     if (DateHelper.isValidDate(visibleValue)) {
+       this.updateValidEventValues(visibleValue);
+     }
+   }
+
   closeDatePicker = () => {
     if (this.isOpening) {
       this.isOpening = false;
       return;
     }
     document.removeEventListener('click', this.closeDatePicker);
-    this.setState({ isDatePickerOpen: false });
+    this.setState({ isDatePickerOpen: false }, () => {
+      this.isBlurBlocked = false;
+      if (this.input !== document.activeElement) {
+        this.handleBlur();
+      }
+    });
   };
 
   handleDateSelect = (selectedDate) => {
     const stringDateIso = DateHelper.formatDateString(selectedDate);
-
     this.isBlurBlocked = true;
-    this.closeDatePicker();
-    this.updateVisibleValue(stringDateIso);
-    this.emitOnChangeCallback(stringDateIso);
+    this.isOpening = false;
+    this.updateVisibleValue(stringDateIso, true);
   };
 
-  updateVisibleValue = (date) => {
+  updateVisibleValue = (date, pickerUsed) => {
     const visibleValue = formatDateToCurrentLocale(date);
+    const newDate = this.getDateObject(date);
 
     this.setState({
-      selectedDate: date,
-      visibleValue
+      selectedDate: newDate
+    },
+    () => {
+      this.updateValidEventValues(visibleValue);
```

Comment:
A couple of lines above this line, it updates the state – `this.setState({ selectedDate: newDate }` – and it passes a callback which waits for the state to update.

This `updateValidEventValues()` call also updates the state – `this.setState({ visibleValue, lastValidEventValues })` – but it doesn't pass a callback, so it doesn't wait for the state to update before proceeding to execute the rest of this function.

Does that matter? It may not matter, but I just wanted to check.

Reply:
In this scenario it is fine, the callback is needed to ensure the selected date is updated, the rest of the stat update does not matter as it has updated accordingly by the time the `onBlur` triggers

Reply:
OK, thanks :ok_hand: :white_check_mark: